dali/operators/reader/numpy_reader_gpu_op.cc

```diff
@@ -0,0 +1,163 @@
+        "[0] has ", ref_shape.size(), " dimensions."));
+    DALI_ENFORCE(ref_shape.size() == sample->shape.size(),
+                  make_string("The tensors in the input batch do not all have the same number of "
+                  "dimensions for sample [0]: ", ref_shape.size(), " vs [", data_idx, "]: ",
+                  sample->shape.size(), "."));
+    tmp_shapes.push_back(sample->shape);
+  }
+
+  curr_tensor_list.Resize(TensorListShape<>(tmp_shapes), ref_type);
+
+  // read the data
+  for (size_t data_idx = 0; data_idx < curr_tensor_list.ntensor(); ++data_idx) {
+    thread_pool_.AddWork([&curr_batch, &curr_tensor_list, data_idx](int tid) {
+        curr_batch[data_idx]->read_sample_f(curr_tensor_list.raw_mutable_data(),
+                                            curr_tensor_list.tensor_offset(data_idx) *
+                                            curr_tensor_list.type().size(),
+                                            curr_tensor_list.nbytes());
+      });
+  }
+  thread_pool_.RunAll();
+}
+
+void PermuteHelper(const TensorShape<> &plain_shapes, std::vector<int64_t> &perm_shape,
+                  std::vector<int> &perm) {
+  int n_dims = plain_shapes.size();
+  if (perm.empty()) {
+    perm.resize(n_dims);
+    for (int i = 0; i < n_dims; ++i) {
+      perm[i] = n_dims - i - 1;
+    }
+  }
+  for (int i = 0; i < n_dims; ++i) {
+    perm_shape[i] = plain_shapes[perm[i]];
+  }
+}
+
+void NumpyReaderGPU::RunImpl(DeviceWorkspace &ws) {
+  TensorListShape<> shape(batch_size_);
+  // use vector for temporarily storing shapes
+  std::vector<TensorShape<>> tmp_shapes;
+  std::vector<TensorShape<>> transpose_shapes;
+  std::vector<int> perm;
+  std::vector<int64_t> perm_shape;
+
+  perm.reserve(GetSampleShape(0).size());
+  perm_shape.resize(GetSampleShape(0).size());
+
+  for (int sample_idx = 0; sample_idx < batch_size_; sample_idx++) {
+    const auto& imfile = GetSample(sample_idx);
+    auto plain_shape = GetSampleShape(sample_idx);
+    if (imfile.meta == "transpose:false") {
+      tmp_shapes.push_back(plain_shape);
+    } else {
+      PermuteHelper(plain_shape, perm_shape, perm);
+      tmp_shapes.push_back(perm_shape);
+      transpose_shapes.push_back(plain_shape);
+    }
+  }
+  auto ref_type = GetSampleType(0);
+  shape = TensorListShape<>(tmp_shapes);
+  ws.Output<GPUBackend>(0).Resize(shape, ref_type);
+
+  auto &image_output = ws.Output<GPUBackend>(0);
+
+  SmallVector<int64_t, 256> copy_sizes;
+  copy_sizes.reserve(batch_size_);
+  SmallVector<const void *, 256> copy_from;
+  copy_from.reserve(batch_size_);
+  SmallVector<void *, 256> copy_to;
+  copy_to.reserve(batch_size_);
+
+  SmallVector<const void *, 256> transpose_from;
+  transpose_from.reserve(batch_size_);
+  SmallVector<void *, 256> transpose_to;
+  transpose_to.reserve(batch_size_);
+
+
+  for (int data_idx = 0; data_idx < batch_size_; ++data_idx) {
+    const auto& imfile = GetSample(data_idx);
+    if (imfile.meta == "transpose:false") {
```

Comment:
really? what was wrong with the boolean?

Reply:
Done